python/paddle/fluid/dygraph/dygraph_to_static/partial_program.py

```diff
@@ -45,6 +45,7 @@ def __init__(self, main_program, inputs, outputs, parameters=None):
         self.outputs = outputs
         self._params = parameters
         self._infer_program = main_program
+        check_params_init_outside(main_program, parameters)
```

Comment:
Not very good ... you have background that ``parameters`` must be ``param_from_init``, but future software engineers may not know it ... then it disobeys OOP: caller just knows how to use interface but doesn't care about background and  implementation details. ``parameters`` can even be ``None`` which can cause problem.

Reply:
Thanks, I refine these code. Actually, `partial_program` only require two conditions:
 - parameter in `self._params` should be type `framework.ParamBase` which are created in dygraph.
- parameter from transformed main_program  shall be found in `self._params`. Because they share same data with `ParamBase` of original dygraph.

`parameters` can be None，which means there is no trainable parameter in `transformed main_program`. I add related checks for this case.